Crash - Parcelising ReactionsSummaryData

```
java.lang.RuntimeException: 
  at android.os.Parcel.writeSerializable (Parcel.java:1833)
  at im.vector.app.features.home.room.detail.timeline.item.ReactionsSummaryData.writeToParcel (Unknown Source:54)
  at im.vector.app.features.home.room.detail.timeline.item.MessageInformationData.writeToParcel (Unknown Source:66)
  at im.vector.app.features.home.room.detail.timeline.action.TimelineEventFragmentArgs.writeToParcel (Unknown Source:17)
  at android.os.Parcel.writeParcelable (Parcel.java:1801)
  at android.os.Parcel.writeValue (Parcel.java:1707)
  at android.os.Parcel.writeArrayMapInternal (Parcel.java:928)
  at android.os.BaseBundle.writeToParcelInner (BaseBundle.java:1588)
  at android.os.Bundle.writeToParcel (Bundle.java:1253)
  at android.os.Parcel.writeBundle (Parcel.java:997)
  at androidx.fragment.app.FragmentState.writeToParcel (FragmentState.java:10)
  at android.os.Parcel.writeTypedObject (Parcel.java:1634)
  at android.os.Parcel.writeTypedList (Parcel.java:1513)
  at android.os.Parcel.writeTypedList (Parcel.java:1470)
  at androidx.fragment.app.FragmentManagerState.writeToParcel (FragmentManagerState.java:1)
  at android.os.Parcel.writeParcelable (Parcel.java:1801)
  at android.os.Parcel.writeValue (Parcel.java:1707)
  at android.os.Parcel.writeArrayMapInternal (Parcel.java:928)
  at android.os.BaseBundle.writeToParcelInner (BaseBundle.java:1588)
  at android.os.Bundle.writeToParcel (Bundle.java:1253)
  at android.os.Parcel.writeBundle (Parcel.java:997)
  at android.os.Parcel.writeValue (Parcel.java:1698)
  at android.os.Parcel.writeArrayMapInternal (Parcel.java:928)
  at android.os.BaseBundle.writeToParcelInner (BaseBundle.java:1588)
  at android.os.Bundle.writeToParcel (Bundle.java:1253)
  at android.os.Parcel.writeBundle (Parcel.java:997)
  at android.os.Parcel.writeValue (Parcel.java:1698)
  at android.os.Parcel.writeArrayMapInternal (Parcel.java:928)
  at android.os.BaseBundle.writeToParcelInner (BaseBundle.java:1588)
  at android.os.Bundle.writeToParcel (Bundle.java:1253)
  at android.app.IActivityTaskManager$Stub$Proxy.activityStopped (IActivityTaskManager.java:4542)
  at android.app.servertransaction.PendingTransactionActions$StopInfo.run (PendingTransactionActions.java:145)
  at android.os.Handler.handleCallback (Handler.java:888)
  at android.os.Handler.dispatchMessage (Handler.java:100)
  at android.os.Looper.loop (Looper.java:213)
  at android.app.ActivityThread.main (ActivityThread.java:8178)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:513)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1101)
Caused by: java.io.NotSerializableException: 
  at java.io.ObjectOutputStream.writeObject0 (ObjectOutputStream.java:1240)
  at java.io.ObjectOutputStream.defaultWriteFields (ObjectOutputStream.java:1604)
  at java.io.ObjectOutputStream.writeSerialData (ObjectOutputStream.java:1565)
  at java.io.ObjectOutputStream.writeOrdinaryObject (ObjectOutputStream.java:1488)
  at java.io.ObjectOutputStream.writeObject0 (ObjectOutputStream.java:1234)
  at java.io.ObjectOutputStream.writeObject (ObjectOutputStream.java:354)
  at android.os.Parcel.writeSerializable (Parcel.java:1828)
```

| Occurrences (30 days) | Total Occurrences | Impacted users (lifetime) |
| --- | --- | --- |
| 660 | 689 | 592 |

https://play.google.com/console/u/0/developers/{redacted}/app/{redacted}/vitals/crashes/99f978d5/details?versionCode=40104002/details?installedFrom=PLAY_STORE&days=30

--- a/changelog.d/5463.bugfix
+++ b/changelog.d/5463.bugfix
@@ -1,0 +1,1 @@
+Fix reactions summary crash

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/TimelineEventController.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/TimelineEventController.kt
@@ -419,7 +419,7 @@ class TimelineEventController @Inject constructor(private val dateFormatter: Vec
                         eventsGroup = timelineEventsGroup,
                         reactionsSummaryEvents = ReactionsSummaryEvents(
                                 onAddMoreClicked = { reactionListFactory.onAddMoreClicked(callback, event) },
-                               onShowLessClicked =  { reactionListFactory.onShowLessClicked(event.eventId) },
+                                onShowLessClicked = { reactionListFactory.onShowLessClicked(event.eventId) },
                                 onShowMoreClicked = { reactionListFactory.onShowMoreClicked(event.eventId) }
                         )
                 )

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/TimelineEventController.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/TimelineEventController.kt
@@ -53,6 +53,7 @@ import im.vector.app.features.home.room.detail.timeline.item.DaySeparatorItem
 import im.vector.app.features.home.room.detail.timeline.item.DaySeparatorItem_
 import im.vector.app.features.home.room.detail.timeline.item.ItemWithEvents
 import im.vector.app.features.home.room.detail.timeline.item.MessageInformationData
+import im.vector.app.features.home.room.detail.timeline.item.ReactionsSummaryEvents
 import im.vector.app.features.home.room.detail.timeline.item.ReadReceiptData
 import im.vector.app.features.home.room.detail.timeline.item.ReadReceiptsItem
 import im.vector.app.features.home.room.detail.timeline.url.PreviewUrlRetriever
@@ -415,7 +416,12 @@ class TimelineEventController @Inject constructor(private val dateFormatter: Vec
                         partialState = partialState,
                         lastSentEventIdWithoutReadReceipts = lastSentEventWithoutReadReceipts,
                         callback = callback,
-                        eventsGroup = timelineEventsGroup
+                        eventsGroup = timelineEventsGroup,
+                        reactionsSummaryEvents = ReactionsSummaryEvents(
+                                onAddMoreClicked = { reactionListFactory.onAddMoreClicked(callback, event) },
+                               onShowLessClicked =  { reactionListFactory.onShowLessClicked(event.eventId) },
+                                onShowMoreClicked = { reactionListFactory.onShowMoreClicked(event.eventId) }
+                        )
                 )
                 modelCache[position] = buildCacheItem(params)
                 numberOfEventsToBuild++

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/CallItemFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/CallItemFactory.kt
@@ -26,6 +26,7 @@ import im.vector.app.features.home.room.detail.timeline.helper.MessageItemAttrib
 import im.vector.app.features.home.room.detail.timeline.item.CallTileTimelineItem
 import im.vector.app.features.home.room.detail.timeline.item.CallTileTimelineItem_
 import im.vector.app.features.home.room.detail.timeline.item.MessageInformationData
+import im.vector.app.features.home.room.detail.timeline.item.ReactionsSummaryEvents
 import org.matrix.android.sdk.api.session.Session
 import org.matrix.android.sdk.api.session.events.model.EventType
 import org.matrix.android.sdk.api.session.room.model.RoomSummary
@@ -61,7 +62,8 @@ class CallItemFactory @Inject constructor(
                             highlight = params.isHighlighted,
                             informationData = informationData,
                             isStillActive = callEventGrouper.isInCall(),
-                            formattedDuration = callEventGrouper.formattedDuration()
+                            formattedDuration = callEventGrouper.formattedDuration(),
+                            reactionsSummaryEvents = params.reactionsSummaryEvents
                     )
                 } else {
                     null
@@ -78,7 +80,8 @@ class CallItemFactory @Inject constructor(
                             highlight = params.isHighlighted,
                             informationData = informationData,
                             isStillActive = callEventGrouper.isRinging(),
-                            formattedDuration = callEventGrouper.formattedDuration()
+                            formattedDuration = callEventGrouper.formattedDuration(),
+                            reactionsSummaryEvents = params.reactionsSummaryEvents
                     )
                 } else {
                     null
@@ -94,7 +97,8 @@ class CallItemFactory @Inject constructor(
                         highlight = params.isHighlighted,
                         informationData = informationData,
                         isStillActive = false,
-                        formattedDuration = callEventGrouper.formattedDuration()
+                        formattedDuration = callEventGrouper.formattedDuration(),
+                        reactionsSummaryEvents = params.reactionsSummaryEvents
                 )
             }
             EventType.CALL_HANGUP -> {
@@ -111,7 +115,8 @@ class CallItemFactory @Inject constructor(
                         highlight = params.isHighlighted,
                         informationData = informationData,
                         isStillActive = false,
-                        formattedDuration = callEventGrouper.formattedDuration()
+                        formattedDuration = callEventGrouper.formattedDuration(),
+                        reactionsSummaryEvents = params.reactionsSummaryEvents
                 )
             }
             else                  -> null
@@ -133,10 +138,11 @@ class CallItemFactory @Inject constructor(
             highlight: Boolean,
             isStillActive: Boolean,
             formattedDuration: String,
-            callback: TimelineEventController.Callback?
+            callback: TimelineEventController.Callback?,
+            reactionsSummaryEvents: ReactionsSummaryEvents?
     ): CallTileTimelineItem? {
         val userOfInterest = roomSummary.toMatrixItem()
-        val attributes = messageItemAttributesFactory.create(null, informationData, callback).let {
+        val attributes = messageItemAttributesFactory.create(null, informationData, callback, reactionsSummaryEvents).let {
             CallTileTimelineItem.Attributes(
                     callId = callId,
                     callKind = callKind,
@@ -151,7 +157,8 @@ class CallItemFactory @Inject constructor(
                     readReceiptsCallback = it.readReceiptsCallback,
                     userOfInterest = userOfInterest,
                     callback = callback,
-                    isStillActive = isStillActive
+                    isStillActive = isStillActive,
+                    reactionsSummaryEvents = reactionsSummaryEvents
             )
         }
         return CallTileTimelineItem_()

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/EncryptedItemFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/EncryptedItemFactory.kt
@@ -111,7 +111,9 @@ class EncryptedItemFactory @Inject constructor(private val messageInformationDat
                         messageContent = event.root.content.toModel<EncryptedEventContent>(),
                         informationData = informationData,
                         callback = params.callback,
-                        threadDetails = threadDetails)
+                        threadDetails = threadDetails,
+                        reactionsSummaryEvents = params.reactionsSummaryEvents
+                )
                 return MessageTextItem_()
                         .layout(informationData.messageLayout.layoutRes)
                         .leftGuideline(avatarSizeProvider.leftGuideline)

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/EncryptionItemFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/EncryptionItemFactory.kt
@@ -46,7 +46,7 @@ class EncryptionItemFactory @Inject constructor(
         }
         val algorithm = event.root.content.toModel<EncryptionEventContent>()?.algorithm
         val informationData = informationDataFactory.create(params)
-        val attributes = messageItemAttributesFactory.create(null, informationData, params.callback)
+        val attributes = messageItemAttributesFactory.create(null, informationData, params.callback, params.reactionsSummaryEvents)
 
         val isSafeAlgorithm = algorithm == MXCRYPTO_ALGORITHM_MEGOLM
         val title: String
@@ -80,7 +80,8 @@ class EncryptionItemFactory @Inject constructor(
                                 itemClickListener = attributes.itemClickListener,
                                 itemLongClickListener = attributes.itemLongClickListener,
                                 reactionPillCallback = attributes.reactionPillCallback,
-                                readReceiptsCallback = attributes.readReceiptsCallback
+                                readReceiptsCallback = attributes.readReceiptsCallback,
+                                reactionsSummaryEvents = attributes.reactionsSummaryEvents
                         )
                 )
                 .highlighted(params.isHighlighted)

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt
@@ -155,7 +155,7 @@ class MessageItemFactory @Inject constructor(
 
         if (event.root.isRedacted()) {
             // message is redacted
-            val attributes = messageItemAttributesFactory.create(null, informationData, callback, threadDetails)
+            val attributes = messageItemAttributesFactory.create(null, informationData, callback, params.reactionsSummaryEvents)
             return buildRedactedItem(attributes, highlight)
         }
 
@@ -177,7 +177,7 @@ class MessageItemFactory @Inject constructor(
         }
 
         // always hide summary when we are on thread timeline
-        val attributes = messageItemAttributesFactory.create(messageContent, informationData, callback, threadDetails)
+        val attributes = messageItemAttributesFactory.create(messageContent, informationData, callback, params.reactionsSummaryEvents, threadDetails)
 
 //        val all = event.root.toContent()
 //        val ev = all.toModel<Event>()
@@ -413,7 +413,8 @@ class MessageItemFactory @Inject constructor(
                                 itemClickListener = attributes.itemClickListener,
                                 reactionPillCallback = attributes.reactionPillCallback,
                                 readReceiptsCallback = attributes.readReceiptsCallback,
-                                emojiTypeFace = attributes.emojiTypeFace
+                                emojiTypeFace = attributes.emojiTypeFace,
+                                reactionsSummaryEvents = attributes.reactionsSummaryEvents
                         )
                 )
                 .callback(callback)

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/TimelineItemFactoryParams.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/TimelineItemFactoryParams.kt
@@ -18,6 +18,7 @@ package im.vector.app.features.home.room.detail.timeline.factory
 
 import im.vector.app.features.home.room.detail.timeline.TimelineEventController
 import im.vector.app.features.home.room.detail.timeline.helper.TimelineEventsGroup
+import im.vector.app.features.home.room.detail.timeline.item.ReactionsSummaryEvents
 import org.matrix.android.sdk.api.session.room.timeline.TimelineEvent
 
 data class TimelineItemFactoryParams(
@@ -29,6 +30,7 @@ data class TimelineItemFactoryParams(
         val partialState: TimelineEventController.PartialState = TimelineEventController.PartialState(),
         val lastSentEventIdWithoutReadReceipts: String? = null,
         val callback: TimelineEventController.Callback? = null,
+        val reactionsSummaryEvents: ReactionsSummaryEvents? = null,
         val eventsGroup: TimelineEventsGroup? = null
 ) {
 

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/VerificationItemFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/VerificationItemFactory.kt
@@ -71,10 +71,10 @@ class VerificationItemFactory @Inject constructor(
         // If it's not a request ignore this event
         // if (refEvent.root.getClearContent().toModel<MessageVerificationRequestContent>() == null) return ignoredConclusion(event, highlight, callback)
 
-        val referenceInformationData = messageInformationDataFactory.create(TimelineItemFactoryParams(refEvent))
+        val referenceInformationData = messageInformationDataFactory.create(TimelineItemFactoryParams(event = refEvent))
 
         val informationData = messageInformationDataFactory.create(params)
-        val attributes = messageItemAttributesFactory.create(null, informationData, params.callback)
+        val attributes = messageItemAttributesFactory.create(null, informationData, params.callback, params.reactionsSummaryEvents)
 
         when (event.root.getClearType()) {
             EventType.KEY_VERIFICATION_CANCEL -> {
@@ -100,7 +100,8 @@ class VerificationItemFactory @Inject constructor(
                                                 itemClickListener = attributes.itemClickListener,
                                                 itemLongClickListener = attributes.itemLongClickListener,
                                                 reactionPillCallback = attributes.reactionPillCallback,
-                                                readReceiptsCallback = attributes.readReceiptsCallback
+                                                readReceiptsCallback = attributes.readReceiptsCallback,
+                                                reactionsSummaryEvents = attributes.reactionsSummaryEvents
                                         )
                                 )
                                 .highlighted(params.isHighlighted)
@@ -133,7 +134,8 @@ class VerificationItemFactory @Inject constructor(
                                         itemClickListener = attributes.itemClickListener,
                                         itemLongClickListener = attributes.itemLongClickListener,
                                         reactionPillCallback = attributes.reactionPillCallback,
-                                        readReceiptsCallback = attributes.readReceiptsCallback
+                                        readReceiptsCallback = attributes.readReceiptsCallback,
+                                        reactionsSummaryEvents = attributes.reactionsSummaryEvents
                                 )
                         )
                         .highlighted(params.isHighlighted)

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/WidgetItemFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/WidgetItemFactory.kt
@@ -88,7 +88,8 @@ class WidgetItemFactory @Inject constructor(
                 userOfInterest = userOfInterest,
                 callback = params.callback,
                 isStillActive = isCallStillActive,
-                formattedDuration = ""
+                formattedDuration = "",
+                reactionsSummaryEvents = params.reactionsSummaryEvents
         )
         return CallTileTimelineItem_()
                 .attributes(attributes)

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/MessageInformationDataFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/MessageInformationDataFactory.kt
@@ -93,7 +93,7 @@ class MessageInformationDataFactory @Inject constructor(private val session: Ses
                 avatarUrl = event.senderInfo.avatarUrl,
                 memberName = event.senderInfo.disambiguatedDisplayName,
                 messageLayout = messageLayout,
-                reactionsSummary = reactionsSummaryFactory.create(event, params.callback),
+                reactionsSummary = reactionsSummaryFactory.create(event),
                 pollResponseAggregatedSummary = event.annotations?.pollResponseSummary?.let {
                     PollResponseData(
                             myVote = it.aggregatedContent?.myVote,

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/MessageItemAttributesFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/MessageItemAttributesFactory.kt
@@ -24,6 +24,7 @@ import im.vector.app.features.home.room.detail.timeline.MessageColorProvider
 import im.vector.app.features.home.room.detail.timeline.TimelineEventController
 import im.vector.app.features.home.room.detail.timeline.item.AbsMessageItem
 import im.vector.app.features.home.room.detail.timeline.item.MessageInformationData
+import im.vector.app.features.home.room.detail.timeline.item.ReactionsSummaryEvents
 import org.matrix.android.sdk.api.session.threads.ThreadDetails
 import javax.inject.Inject
 
@@ -38,6 +39,7 @@ class MessageItemAttributesFactory @Inject constructor(
     fun create(messageContent: Any?,
                informationData: MessageInformationData,
                callback: TimelineEventController.Callback?,
+               reactionsSummaryEvents: ReactionsSummaryEvents?,
                threadDetails: ThreadDetails? = null): AbsMessageItem.Attributes {
         return AbsMessageItem.Attributes(
                 avatarSize = avatarSizeProvider.avatarSize,
@@ -60,6 +62,7 @@ class MessageItemAttributesFactory @Inject constructor(
                 emojiTypeFace = emojiCompatFontProvider.typeface,
                 decryptionErrorMessage = stringProvider.getString(R.string.encrypted_message),
                 threadDetails = threadDetails,
+                reactionsSummaryEvents = reactionsSummaryEvents,
                 areThreadMessagesEnabled = preferencesProvider.areThreadMessagesEnabled()
         )
     }

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/ReactionsSummaryFactory.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/ReactionsSummaryFactory.kt
@@ -34,7 +34,7 @@ class ReactionsSummaryFactory @Inject constructor() {
         return eventsRequestingBuild.remove(event.eventId)
     }
 
-    fun create(event: TimelineEvent, callback: TimelineEventController.Callback?): ReactionsSummaryData {
+    fun create(event: TimelineEvent): ReactionsSummaryData {
         val eventId = event.eventId
         val showAllStates = showAllReactionsByEvent.contains(eventId)
         val reactions = event.annotations?.reactionsSummary
@@ -43,19 +43,22 @@ class ReactionsSummaryFactory @Inject constructor() {
                 }
         return ReactionsSummaryData(
                 reactions = reactions,
-                showAll = showAllStates,
-                onShowMoreClicked = {
-                    showAllReactionsByEvent.add(eventId)
-                    onRequestBuild(eventId)
-                },
-                onShowLessClicked = {
-                    showAllReactionsByEvent.remove(eventId)
-                    onRequestBuild(eventId)
-                },
-                onAddMoreClicked = {
-                    callback?.onAddMoreReaction(event)
-                }
+                showAll = showAllStates
         )
+    }
+
+    fun onAddMoreClicked(callback: TimelineEventController.Callback?, event: TimelineEvent) {
+        callback?.onAddMoreReaction(event)
+    }
+
+    fun onShowMoreClicked(eventId: String) {
+        showAllReactionsByEvent.add(eventId)
+        onRequestBuild(eventId)
+    }
+
+    fun onShowLessClicked(eventId: String) {
+        showAllReactionsByEvent.remove(eventId)
+        onRequestBuild(eventId)
     }
 
     private fun onRequestBuild(eventId: String) {

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsBaseMessageItem.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsBaseMessageItem.kt
@@ -121,18 +121,24 @@ abstract class AbsBaseMessageItem<H : AbsBaseMessageItem.Holder> : BaseEventItem
                 val showReactionsTextView = createReactionTextView(holder)
                 if (reactionsSummary.showAll) {
                     showReactionsTextView.setText(R.string.message_reaction_show_less)
-                    showReactionsTextView.onClick { reactionsSummary.onShowLessClicked() }
+                    showReactionsTextView.onClick {
+                        baseAttributes.reactionsSummaryEvents?.onShowLessClicked?.invoke()
+                    }
                 } else {
                     val moreCount = reactions.count() - MAX_REACTIONS_TO_SHOW
                     showReactionsTextView.text = holder.view.resources.getQuantityString(R.plurals.message_reaction_show_more, moreCount, moreCount)
-                    showReactionsTextView.onClick { reactionsSummary.onShowMoreClicked() }
+                    showReactionsTextView.onClick {
+                        baseAttributes.reactionsSummaryEvents?.onShowMoreClicked?.invoke()
+                    }
                 }
                 holder.reactionsContainer.addView(showReactionsTextView)
             }
             val addMoreReactionsTextView = createReactionTextView(holder)
 
             addMoreReactionsTextView.text = holder.view.context.getDrawableAsSpannable(R.drawable.ic_add_reaction_small)
-            addMoreReactionsTextView.onClick { reactionsSummary.onAddMoreClicked() }
+            addMoreReactionsTextView.onClick {
+                baseAttributes.reactionsSummaryEvents?.onAddMoreClicked?.invoke()
+            }
             holder.reactionsContainer.addView(addMoreReactionsTextView)
             holder.reactionsContainer.setOnLongClickListener(baseAttributes.itemLongClickListener)
         }
@@ -180,6 +186,7 @@ abstract class AbsBaseMessageItem<H : AbsBaseMessageItem.Holder> : BaseEventItem
 
         //        val memberClickListener: ClickListener?
         val reactionPillCallback: TimelineEventController.ReactionPillCallback?
+        val reactionsSummaryEvents: ReactionsSummaryEvents?
 
         //        val avatarCallback: TimelineEventController.AvatarCallback?
         val readReceiptsCallback: TimelineEventController.ReadReceiptsCallback?

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsMessageItem.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/AbsMessageItem.kt
@@ -184,7 +184,8 @@ abstract class AbsMessageItem<H : AbsMessageItem.Holder> : AbsBaseMessageItem<H>
             val emojiTypeFace: Typeface? = null,
             val decryptionErrorMessage: String? = null,
             val threadDetails: ThreadDetails? = null,
-            val areThreadMessagesEnabled: Boolean = false
+            val areThreadMessagesEnabled: Boolean = false,
+            override val reactionsSummaryEvents: ReactionsSummaryEvents? = null,
     ) : AbsBaseMessageItem.Attributes {
 
         // Have to override as it's used to diff epoxy items

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/CallTileTimelineItem.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/CallTileTimelineItem.kt
@@ -263,7 +263,8 @@ abstract class CallTileTimelineItem : AbsBaseMessageItem<CallTileTimelineItem.Ho
             override val itemLongClickListener: View.OnLongClickListener? = null,
             override val itemClickListener: ClickListener? = null,
             override val reactionPillCallback: TimelineEventController.ReactionPillCallback? = null,
-            override val readReceiptsCallback: TimelineEventController.ReadReceiptsCallback? = null
+            override val readReceiptsCallback: TimelineEventController.ReadReceiptsCallback? = null,
+            override val reactionsSummaryEvents: ReactionsSummaryEvents? = null
     ) : AbsBaseMessageItem.Attributes
 
     enum class CallKind(@DrawableRes val icon: Int, @StringRes val title: Int) {

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/MessageInformationData.kt
@@ -58,11 +58,14 @@ data class ReferencesInfoData(
 data class ReactionsSummaryData(
         /*List of reactions (emoji,count,isSelected)*/
         val reactions: List<ReactionInfoData>? = null,
-        val showAll: Boolean = false,
+        val showAll: Boolean = false
+) : Parcelable
+
+data class ReactionsSummaryEvents(
         val onShowMoreClicked: () -> Unit,
         val onShowLessClicked: () -> Unit,
         val onAddMoreClicked: () -> Unit
-) : Parcelable
+)
 
 @Parcelize
 data class ReactionInfoData(

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/StatusTileTimelineItem.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/StatusTileTimelineItem.kt
@@ -93,7 +93,8 @@ abstract class StatusTileTimelineItem : AbsBaseMessageItem<StatusTileTimelineIte
             override val itemClickListener: ClickListener? = null,
             override val reactionPillCallback: TimelineEventController.ReactionPillCallback? = null,
             override val readReceiptsCallback: TimelineEventController.ReadReceiptsCallback? = null,
-            val emojiTypeFace: Typeface? = null
+            val emojiTypeFace: Typeface? = null,
+            override val reactionsSummaryEvents: ReactionsSummaryEvents? = null
     ) : AbsBaseMessageItem.Attributes
 
     enum class ShieldUIState {

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/VerificationRequestItem.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/VerificationRequestItem.kt
@@ -154,6 +154,7 @@ abstract class VerificationRequestItem : AbsBaseMessageItem<VerificationRequestI
             override val reactionPillCallback: TimelineEventController.ReactionPillCallback? = null,
 //            val avatarCallback: TimelineEventController.AvatarCallback? = null,
             override val readReceiptsCallback: TimelineEventController.ReadReceiptsCallback? = null,
-            val emojiTypeFace: Typeface? = null
+            override val reactionsSummaryEvents: ReactionsSummaryEvents? = null,
+            val emojiTypeFace: Typeface? = null,
     ) : AbsBaseMessageItem.Attributes
 }

--- a/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/WidgetTileTimelineItem.kt
+++ b/vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/WidgetTileTimelineItem.kt
@@ -83,6 +83,7 @@ abstract class WidgetTileTimelineItem : AbsBaseMessageItem<WidgetTileTimelineIte
             override val itemClickListener: ClickListener? = null,
             override val reactionPillCallback: TimelineEventController.ReactionPillCallback? = null,
             override val readReceiptsCallback: TimelineEventController.ReadReceiptsCallback? = null,
-            val emojiTypeFace: Typeface? = null
+            val emojiTypeFace: Typeface? = null,
+            override val reactionsSummaryEvents: ReactionsSummaryEvents? = null
     ) : AbsBaseMessageItem.Attributes
 }

